tests/client/test_connector.py

```diff
@@ -1,206 +1,396 @@
+        nonlocal server_channel_transport, transport_creation_calls
+        transport_creation_calls += 1
+        assert transport_creation_calls == 1
+        server_channel_transport = ChannelTransport(channel)
+        server_channel_transport.start_reader()
+        return server_channel_transport
+
+    with patch("snitun.client.connector.ChannelTransport", _save_transport):
+        response = await session.get("https://localhost:4242/")
+        assert response.status == 200
+        content = await response.read()
+        assert content == b"Hello world"
+
+    # Simulate an exception getting the buffer
+    assert server_channel_transport is not None
+    ssl_proto = cast(
+        asyncio.sslproto.SSLProtocol,
+        server_channel_transport.get_protocol(),
+    )
+    assert server_channel_transport.is_reading
+    with patch.object(ssl_proto, "get_buffer", side_effect=exception):
+        response = await session.get("https://localhost:4242/")
+        assert response.status == 200
+        content = await response.read()
+        assert content == b"Hello world"
+
+    await session.close()
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 11),
+    reason="Requires Python 3.11+ for working start_tls",
+)
+@pytest.mark.parametrize("exception", [MultiplexerTransportClose, Exception])
+async def test_connector_valid_url_buffer_updated_raises(
+    multiplexer_client: Multiplexer,
+    multiplexer_server: Multiplexer,
+    connector: Connector,
+    client_ssl_context: ssl.SSLContext,
+    exception: Exception,
+) -> None:
+    """End to end test that connector fails when updating the buffer raises."""
+    multiplexer_client._new_connections = connector.handler
+    connector = ChannelConnector(multiplexer_server, client_ssl_context)
+    session = aiohttp.ClientSession(connector=connector)
+
+    server_channel_transport: ChannelTransport | None = None
+    transport_creation_calls = 0
+
+    def _save_transport(channel: MultiplexerChannel) -> ChannelTransport:
+        nonlocal server_channel_transport, transport_creation_calls
+        transport_creation_calls += 1
+        assert transport_creation_calls == 1
+        server_channel_transport = ChannelTransport(channel)
+        server_channel_transport.start_reader()
+        return server_channel_transport
+
+    with patch("snitun.client.connector.ChannelTransport", _save_transport):
+        response = await session.get("https://localhost:4242/")
+        assert response.status == 200
+        content = await response.read()
+        assert content == b"Hello world"
+
+    # Simulate an exception getting the buffer
+    assert server_channel_transport is not None
+    ssl_proto = cast(
+        asyncio.sslproto.SSLProtocol,
+        server_channel_transport.get_protocol(),
+    )
+    assert server_channel_transport.is_reading
+    with patch.object(ssl_proto, "buffer_updated", side_effect=exception):
+        response = await session.get("https://localhost:4242/")
```

Comment:
Seems like we want this to fail, and it does on the backend but doesn't fail here